pkg/crosscluster/logical/offline_initial_scan_processor.go

```diff
@@ -0,0 +1,447 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package logical
+
+import (
+	"context"
+	"fmt"
+	"runtime/pprof"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/backup"
+	"github.com/cockroachdb/cockroach/pkg/crosscluster"
+	"github.com/cockroachdb/cockroach/pkg/crosscluster/streamclient"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/kv/bulk"
+	"github.com/cockroachdb/cockroach/pkg/repstream/streampb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowexec"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/log/logcrash"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/span"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/logtags"
+)
+
+// TODO(msbutler): we can remove this setting once we eagerly spin down the proc
+// when the initial scan completes.
+var minimumFlushInterval = settings.RegisterDurationSettingWithExplicitUnit(
+	settings.SystemOnly,
+	"logical_replication.offline_scan_flush_interval",
+	"the minimum timestamp between flushes; flushes may still occur if internal buffers fill up",
+	5*time.Second,
+)
+
+// offlineInitialScanProcessor consumes a cross-cluster replication stream
+// by decoding kvs in it to logical changes and applying them by executing DMLs.
+type offlineInitialScanProcessor struct {
```

Comment:
Is it possible to use the PCR processor here instead of creating a processor for an offline initial scan? I'm thinking something like creating the processor with `now()` as the cutover time. Then start LDR.

The PCR processor has a bunch of logic built in for handling things like splits and has more optimizations for getting the initial writes into L6.

Reply:
I actually argued the opposite: the processor has very little in common since we don’t need  buffering and sorting of out of order keys, time-based flushes, reactive splits, etc

Reply:
So, my initial prototype used the existing PCR processor, but I realized it had a few too many bells and whistles that only applied to PCR and not LDR initial scan specifically (e.g. del range ops, buffering writes outside of the batcher). In particular, modifying the processor shutdown logic for LDR initial scan overcomplicated the PCR processor to the point that it seemed easier to start from scratch.